File: core/rule_of_five.py
"""
Rule of 5 — cap list outputs to top 5 brokers/items for optimal LLM signal-to-noise.
See docs/ARCHITECTURE.md §3.
"""

RULE_OF_FIVE = 5


def clamp_limit(limit=None, default_limit=RULE_OF_FIVE):
    if limit is None:
        return default_limit
    try:
        n = int(limit)
    except (TypeError, ValueError):
        return default_limit
    if n < 1:
        return default_limit
    return min(n, RULE_OF_FIVE)


def trim_to_rule_of_five(items, limit=RULE_OF_FIVE):
    if not isinstance(items, list):
        return items
    return items[:limit]

File: core/rule-of-five.js
/**
 * Rule of 5 — cap list outputs to top 5 brokers/items for optimal LLM signal-to-noise.
 * See docs/ARCHITECTURE.md §3.
 */

const RULE_OF_FIVE = 5;

function clampLimit(limit, defaultLimit = RULE_OF_FIVE) {
  if (limit == null) return defaultLimit;
  const n = Number(limit);
  if (Number.isNaN(n) || n < 1) return defaultLimit;
  return Math.min(n, RULE_OF_FIVE);
}

function trimToRuleOfFive(items, limit = RULE_OF_FIVE) {
  if (!Array.isArray(items)) return items;
  return items.slice(0, limit);
}

module.exports = { RULE_OF_FIVE, clampLimit, trimToRuleOfFive };